checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java

```diff
@@ -104,28 +142,353 @@ public LockTransfer createFlowTransferFunction(CFAbstractAnalysis<CFValue, LockS
+            this.annotation = annotation;
+            this.annotationClass = annotationClass;
+        }
+
+        public String getNameOfSideEffectAnnotation() {
+            return annotation;
+        }
+
+        public Class<? extends Annotation> getAnnotationClass() {
+            return annotationClass;
+        }
 
-        AnnotatedTypeMirror type = getAnnotatedType(tree);
+        /**
+         * Returns true if the receiver side effect annotation is weaker
+         * than side effect annotation 'other'.
+         */
+        boolean isWeakerThan(SideEffectAnnotation other) {
+            boolean weaker = false;
 
-        shouldCache = true;
+            switch (other) {
+                case MAYRELEASELOCKS:
+                    break;
+                case RELEASESNOLOCKS:
+                    if (this == SideEffectAnnotation.MAYRELEASELOCKS) {
+                        weaker = true;
+                    }
+                    break;
+                case LOCKINGFREE:
+                    switch (this) {
+                        case MAYRELEASELOCKS:
+                        case RELEASESNOLOCKS:
+                            weaker = true;
+                            break;
+                        default:
+                    }
+                    break;
+                case SIDEEFFECTFREE:
+                    switch (this) {
+                        case MAYRELEASELOCKS:
+                        case RELEASESNOLOCKS:
+                        case LOCKINGFREE:
+                            weaker = true;
+                            break;
+                        default:
+                    }
+                    break;
+                case PURE:
+                    switch (this) {
+                        case MAYRELEASELOCKS:
+                        case RELEASESNOLOCKS:
+                        case LOCKINGFREE:
+                        case SIDEEFFECTFREE:
+                            weaker = true;
+                            break;
+                        default:
+                    }
+                    break;
+            }
 
-        return type;
+            return weaker;
+        }
+
+        static SideEffectAnnotation weakest = null;
+        public static SideEffectAnnotation weakest() {
+            if (weakest == null) {
+                for (SideEffectAnnotation sea : SideEffectAnnotation.values()) {
+                    if (weakest == null) {
+                        weakest = sea;
+                    }
+                    if (sea.isWeakerThan(weakest)) {
+                        weakest = sea;
+                    }
+                }
+            }
+            return weakest;
+        }
     }
 
-    /**
-     * @return The list of annotations of the lock type system.
-     */
-    public Set<Class<? extends Annotation>> getLockAnnotations() {
-        return lockAnnos;
+    // Indicates which side effect annotation is present on the given method.
```

Comment:
If you make this a Javadoc comment IDE will display it nicely when asked at call sites.